File: Project/code/evolve_normal.m
% evolve_normal.m   04/25/2006  Maged Ismail
function evolve_normal

% Interface motion with a constant normal speed for the propagating cosine
% curve test problem
%
% reproducing Fig. 6 (a) of Sun, Y., and Beckermann, C.,
% "Sharp Interface Tracking Using the Phase-Field Equation,"
% J. Computational Physics, Vol. 220, 2007, pp. 626-653.
%
% Finite-difference method is used to solve the phase-field equation.
% Simple forward Euler method (explicit) is used for time discretization. 
%
% User m functions called: distance_function.m, curvature.m, normcd.m and
% laplacian.m

clear all
clc

% Generate the grid and define the parameters
nx=101; ny=nx;
h=1/(nx-1);
W=2*h;
bp=0.5;
b=W*bp; 
W2_inv=1/W^2;
xm=-0.1:h:1.1; ym=-0.1:h:1.1;
[Y,X]=meshgrid(ym,xm);
x=-0.1:h:1.1; y=(1+cos(2*pi*(1-x)))/4;
C=[x;y]';   % contour
Nx=length(xm); Ny=length(ym);
Mark = zeros(size(X));
for i=1:length(xm)
    for j=1:length(ym)
        if (Y(i,j))<(((1+cos(2*pi*(1-X(i,j))))/4))
            Mark(i,j)=-1;
        end
    end
end

for i=1:length(xm)
    for j=1:length(ym)
        phi(i,j)=distance_function(X(i,j),Y(i,j),C);
        if Mark(i,j)==-1
            phi(i,j) = -phi(i,j);
        end
    end
end

for i=1:length(xm)
    for j=1:length(ym)
        if phi(i,j)>0.03
            phi(i,j)=1;
        elseif phi(i,j)<-.03
            phi(i,j)=-1;
        end
    end
end

kappa= curvature(phi,h); normphi=normcd(phi,h); d2phi=laplacian(phi,h);

contour(X,Y,phi,[0 0],'b','linewidth',1);
hold on

dt=0.001; tf=0.1;
t0=0:0.1:0.3; tf=0.1:0.1:0.4;
tic
for p=1:length(tf)
    for t=t0(p):dt:tf(p)
        kappa=curvature(phi,h); normphi=normcd(phi,h); d2phi=laplacian(phi,h);
        for i=2:Nx-1
            for j=2:Ny-1

                phi(i,j)=phi(i,j)+dt*(b*(d2phi(i,j)+W2_inv*phi(i,j)*...
                    (1-phi(i,j)^2)-normphi(i,j)*kappa(i,j))-normphi(i,j));

            end
        end
    end
    contour(X,Y,phi,[0 0],'b','linewidth',1);
end
toc
axis([0 1 -.1 1.1]); xlabel('x'); ylabel('y'); axis square;
% title('Calculated \phi = 0 contours for the propagating cosine curve test problem using a 100 x 120 mesh')
hold off
% end of evolve_normal.m

File: Project/python_convert/evolve_normal.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import sparse
from scipy.sparse.linalg import spsolve
from scipy import integrate
from scipy import interpolate
import time

#Generate the grid and define the parameters
nx=101; ny=nx;
h=1/(nx-1);
W=2*h;
bp=0.5;
b=W*bp; 
W2_inv=1/W**2;
xm=np.arange(-0.1,1.1+h,h); ym=np.arange(-0.1,1.1+h,h);
[X,Y]=np.meshgrid(xm,ym);
x=np.arange(-0.1,1.1+h,h); y=(1+np.cos(2*np.pi*(1-x)))/4;
C=np.vstack((x,y)).T;   # contour
Nx=len(xm); Ny=len(ym);
Mark = np.zeros(X.shape)
for i in range(0,len(xm)):
    for j in range(0,len(ym)):
        if (Y[i,j])<(((1+np.cos(2*np.pi*(1-X[i,j])))/4)):
            Mark[i,j]=-1;
for i in range(0,len(xm)):
    for j in range(0,len(ym)):
        phi[i,j]=distance_function(X[i,j],Y[i,j],C);
        if Mark[i,j]==-1:
            phi[i,j] = -phi[i,j];
for i in range(0,len(xm)):
    for j in range(0,len(ym)):
        if phi[i,j]>0.03:
            phi[i,j]=1;
        elif phi[i,j]<-.03:
            phi[i,j]=-1;

kappa= curvature(phi,h); normphi=normcd(phi,h); d2phi=laplacian(phi,h);

plt.contour(X,Y,phi,[0,0],'b',linewidth=1);
plt.hold(True)

dt=0.001; tf=0.1;
t0=np.arange(0,0.3+0.1,0.1); tf=np.arange(0.1,0.4+0.1,0.1);
tic=time.time()
for p in range(0,len(tf)):
    for t in np.arange(t0[p],tf[p]+dt,dt):
        kappa=curvature(phi,h)
        normphi=normcd(phi,h)
        d2phi=laplacian(phi,h)
        for i in range(1,Nx):
            for j in range(1,Ny):
                phi[i,j]=phi[i,j]+dt*(b*(d2phi[i,j]+W2_inv*phi[i,j]*(1-phi[i,j]**2)-normphi[i,j]*kappa[i,j])-normphi[i,j])